### ISSUE TITLE
[datetime format] Time Zone handling

### ISSUE THREAD

**Issue by fe51** [pyronear/pyro-api#91]
Our data are supposed to be stored as UTC datetime

However, this information is implicit and we have no constraint on it.

What do you think of having datetime stored with time zone information ?

As an example, this format could be convenient "2020-04-01T12:34:56.000+0100".

That being said,to facilitate the use of our api, would be better to store, or at least, return all datetime in "UTC" (ie +0000).  Thus, even if one user, get some that that could have been uploaded with differnt timezone, its gets all data with the same tz info.

So, once we agree on the following points, we can open PR and work on it :)

- [ ] defining the datetime format (maybe we could use timestamptz datatype from postgreSQL ?, but I am not sure what is the easly compliant with sql alchemy)

- [ ]  deciding if we store all datetime with the same time zone information (always in UTC, ie +0000)

- [ ] returning all datetime with the same time zone info (always in UTC, is +0000)  (one day, we might add if needed an option to get data in a specific TZ or to infer the expected tz from the location .. )

Looking for feedbacks :)

**Comment by frgfm** [comment]
Thanks for opening this!
I'm not very knowledgeable about string formatting of datetime, but here are my thoughts:

- the misalignment of datetime can only happen when the input is different (i.e. automatic in the API, or passed by the user)
- in that regards, the event & installation tables might have some consistency issues
- as a first fix, I'd suggest specifying in the doc of both the API & the client that datetime needs to be passed as UTC

On the longterm, we can either store all datetime in UTC or Paris time, but considering the small difference, keeping UTC might be a good idea (especially if we extend beyond French borders). But returning the datetime, perhaps we can add the local time feature only on the client side (pick the IP address, guess the Timezone or add a method argument to pass the TZ).

**Comment by frgfm** [comment]
Hey everyone :wave: 

Let's deal with this issue!
Could you provide some feedback on my previous suggestion? Local time is really a feature specific to the client side. On the API, it's easier to deal with UTC time. So I guess that we should adapt the documentation, and later on, if required, add a local time resolution on the client side. What do you think?

**Comment by jeanpasquier75** [comment]
Good question, what are the use cases of having a user filling a timestamp form ?

IMO it would be easier to store all timestamps in UTC & client handles conversion UTC <-> local time

Some feedbacks: https://dev.to/corykeane/3-simple-rules-for-effectively-handling-dates-and-timezones-1pe0
TLDR:
1. STORE DATETIMES IN UTC IN YOUR DATABASE, AND BACK END CODE
2. CONVERT DATETIMES TO THE USER'S LOCAL TIMEZONE USING FRONTEND CODE

**Comment by frgfm** [comment]
Fully agree with you @jeanpasquier75!
So should we update the documentation, and add a utility on the client side for TZ conversion?

**Comment by florianriche** [comment]
I +1 your suggestion too; let's alter the Client code.

**Comment by fe51** [comment]
Agree with you as well, this is a best practice and easier to handle UTC datetime in backend
Just to be sure, 
- **Client** : we alter the client code to convert datetime to UTC before updating the db
- **Doc**: we explicitly quote that datetime in db are store in UTC and as well in client doc, client will return data in UTC
- **DB datetime format** : we can keep the current format. But do we want to have a format with tz info like "2020-04-01T12:34:56.000+0000" (redundant because always in UTC but at least information about tz is explicit) ?

**Comment by frgfm** [comment]
Respectively:
> Client : we alter the client code to convert datetime to UTC before updating the db
Not sure what you mean here, the only time an actual user will interact with a datetime is for reading access (timestamp at creation is generated automatically), or some specific users to update some tables. For reading, the conversion UTC --> local, will be done entirely on the client and solely for reading (adding a boolean option `convert_to_tz` in corresponding methods?). For updating, a datetime has a timezone linked to it. So perhaps we only need to specify in the doc that authorized users are to pass the datetime as UTC? (I think it's more dangerous to force conversion local --> UTC, depending on the expert-level of the user)

> Doc: we explicitly quote that datetime in db are store in UTC and as well in client doc, client will return data in UTC
cf. my suggestion above regarding the TZ returned on the client side

> DB datetime format : we can keep the current format. But do we want to have a format with tz info like "2020-04-01T12:34:56.000+0000" (redundant because always in UTC but at least information about tz is explicit) ?
Mmmh I have to check this, I'm not an expert on this topic :man_shrugging:

**Comment by frgfm** [comment]
Coming back to the datetime format, I suggest keeping it simple. In the DB, everything is expected to be in UTC, so not TZ in there. Minimal change on the back-end, just a few adjustments on the client side :pray:

**Comment by frgfm** [comment]
Closing this since the timestamp is created by the API in UTC, so there is no way we could have a problem. The only issue that could arise in the future is if the client is able to override or pass timestamps, but this isn't the case right now :)
